File: sumOfDepthofNodes/sumOfDepthofNodes.js
//THE PROMPT
//GIVEN A BINARY TREE, RETURN THE SUM OF THE DEPTH OF EACH NODE. 


//MY ANSWER
// O(N) TIME | O(H) SPACE WHERE H IS THE HEIGHT OF THE TREE
function nodeDepths(root, depth = 0) {
    // Write your code here.
  
      if (root === null) return 0;
      return depth + nodeDepths(root.left, depth + 1) + nodeDepths(root.right, depth + 1)
  }
  
  // This is the class of the input binary tree.
  class BinaryTree {
    constructor(value) {
      this.value = value;
      this.left = null;
      this.right = null;
    }
  }
  

File: sumOfDepthofNodes/sumOfDepthofNodes.java
import java.util.*;

class Program {

  public static int nodeDepths(BinaryTree root) {
    // Write your code here.
		return nodeDepthsHelper(root, 0);	
  }

	public static int nodeDepthsHelper(BinaryTree root, int depth) {
		if (root == null) return 0;
		return depth + nodeDepthsHelper(root.left, depth + 1) + nodeDepthsHelper(root.right, depth + 1);

	}
  static class BinaryTree {
    int value;
    BinaryTree left;
    BinaryTree right;

    public BinaryTree(int value) {
      this.value = value;
      left = null;
      right = null;
    }
  }
}